packages/wallet/src/KeyManagement/util/mapHardwareSigningData.ts

```diff
@@ -0,0 +1,547 @@
+            }
+          : {
+              params: { rewardAccountHex },
+              type: PoolRewardAccountType.THIRD_PARTY
+            };
+      const vrfKeyHashHex = Buffer.from(params.vrf_keyhash().to_bytes()).toString('hex');
+
+      certificate.params = {
+        cost,
+        margin: {
+          denominator: margin.denominator().to_str(),
+          numerator: margin.numerator().to_str()
+        },
+        metadata,
+        pledge,
+        poolKey,
+        poolOwners,
+        relays: ledgerRelays,
+        rewardAccount,
+        vrfKeyHashHex
+      };
+    }
+    certs.push(certificate);
+  }
+  return {
+    certs,
+    signingMode
+  };
+};
+
+const prepareLedgerWithdrawals = (withdrawals: CSL.Withdrawals, rewardAccountKeyPath: BIP32Path): Withdrawal[] => {
+  const ledgerWithdrawals = [];
+  for (let i = 0; i < withdrawals.keys().len(); i++) {
+    const withdrawal = { stakeCredential: {} } as Withdrawal;
+    const rewardAddress = withdrawals.keys().get(i);
+    const paymentCredentials = rewardAddress.payment_cred();
+    const paymentCredentialsScriptHash = paymentCredentials.to_scripthash();
+    if (rewardAddress.payment_cred().kind() === 0) {
+      withdrawal.stakeCredential.type = StakeCredentialParamsType.KEY_PATH;
+      // Ignore! KeyStakeCredentialParams / ScriptStakeCredentialParams are not exported
+      // @ts-ignore
+      withdrawal.stakeCredential.keyPath = rewardAccountKeyPath;
+    } else if (paymentCredentialsScriptHash) {
+      withdrawal.stakeCredential.type = StakeCredentialParamsType.SCRIPT_HASH;
+      // @ts-ignore
+      withdrawal.stakeCredential.scriptHash = Buffer.from(paymentCredentialsScriptHash.to_bytes()).toString('hex');
+    }
+    const withdrawalAmount = withdrawals.get(rewardAddress);
+    if (!withdrawalAmount) {
+      throw new HwMappingError('Withdrawal amount is not defined.');
+    }
+    ledgerWithdrawals.push({
+      ...withdrawal,
+      amount: withdrawalAmount.to_str()
+    });
+  }
+  return ledgerWithdrawals;
+};
+
+const prepareLedgerMintBundle = (
+  mint: CSL.Mint,
+  paymentKeyPaths: BIP32Path[],
+  rewardAccountKeyPath: BIP32Path
+): LedgerMintBundle => {
+  const additionalWitnessPaths: BIP32Path[] = [];
+  const mintAssetsGroup = [];
+
+  for (let j = 0; j < mint.keys().len(); j++) {
+    const policy = mint.keys().get(j);
+    const assets = mint.get(policy);
+    const tokens = [];
+    if (assets) {
+      for (let k = 0; k < assets.keys().len(); k++) {
+        const assetName = assets.keys().get(k);
+        const amount = assets.get(assetName);
+        const positiveAmount = amount?.as_positive()?.to_str();
+        const negativeAmount = amount?.as_negative()?.to_str();
+        if (!amount || !positiveAmount || !negativeAmount) {
+          throw new HwMappingError('Missing token amount.');
+        }
+        tokens.push({
+          amount: amount.is_positive() ? positiveAmount : negativeAmount,
```

Comment:
I'm like 80% sure this is a bug
```suggestion
          amount: amount.is_positive() ? positiveAmount : ('-' + negativeAmount),
```

Reply:
yep, you are right. It should be like here:
https://github.com/input-output-hk/cardano-js-sdk/blob/master/packages/core/src/CSL/cslToCore/cslToCore.ts#L113